Add function browser to the custom expression editor

romeovs opened the pull request:
Closes QUE-583

Adds the function browser to the custom expression editor.

### To verify
1. New question > Orders > Custom column
2. Click the `F` button in the top-right
3. The function browser should appear, try searching it and clicking a function
4. The function should be inserted where your cursor was, with the cursor between its `()`
5. Play around with different screen sizes with the browser open or closed
6. At 640px the whole editor should go full screen

<img width="1243" alt="Screenshot 2025-03-13 at 15 50 52" src="https://github.com/user-attachments/assets/b7022473-41aa-42f8-8c9d-3feaaf0af6ef" />
<img width="1230" alt="Screenshot 2025-03-13 at 15 50 48" src="https://github.com/user-attachments/assets/2897c5c0-8684-457a-a48c-5c13a7c6e795" />
<img width="572" alt="Screenshot 2025-03-13 at 15 52 07" src="https://github.com/user-attachments/assets/87813773-ab7b-4109-9b05-5b7a27d79d0c" />

romeovs committed:
Add FunctionBrowser component

romeovs committed:
Move header into CombineColumns for consistency

romeovs committed:
Render the function browser

romeovs committed:
Allow wrapping shortcuts

romeovs committed:
Render toolbar buttons on top again

romeovs committed:
Add missing Lib import

romeovs committed:
Fix e2e test

romeovs committed:
Use ellipsis in placeholder

romeovs committed:
Add test-id to function browser

romeovs committed:
Add e2e tests for function browser

romeovs committed:
Add category to help text configs

romeovs committed:
Only render supported functions in function browser

romeovs committed:
Convert to helptext in getFilteredClauses

romeovs committed:
Group functions by category

romeovs committed:
More e2e tests

romeovs committed:
Add empty state to function browser search

romeovs committed:
Add more breakpoints for expression editor layout

romeovs committed:
Move function browser item into its own component

romeovs committed:
Merge branch 'master' into custom-expression-function-browser

mngr commented:
I like the behavior! Inserting the function with click is 👌
For some reason case function is not searchable.
<img width="200" alt="Captura de ecrã 2025-03-14, às 15 44 11" src="https://github.com/user-attachments/assets/1bbfcb18-b39a-4a80-b6fc-377816b63d72" />
The buttons for functions and auto-format don't match
<img width="200" alt="Captura de ecrã 2025-03-14, às 15 44 43" src="https://github.com/user-attachments/assets/31836947-68b5-4e32-bdbf-3fd81af39611" />
And the text in the function browser is difficult to read, not sure why
<img width="967" alt="Captura de ecrã 2025-03-14, às 15 43 34" src="https://github.com/user-attachments/assets/9d60ab7a-9fe3-430f-87bd-522c5c78f632" />
@AlessioLaiso any ideas how to make it more legible?

AlessioLaiso commented:
The function browser is much narrower than in the designs, where it's 424px. This should also help making the sentences more readable.

<img width="437" alt="Screenshot 2025-03-14 at 20 50 53" src="https://github.com/user-attachments/assets/046300ca-8830-44c7-a61e-ef55eab09d17" />

romeovs committed:
Increase width of function browser

romeovs committed:
Add case to functions

romeovs commented:
> The buttons for functions and auto-format don't match
Is is as per design. The function browser button is a toggle so it stays blue when it's open.


> The function browser is much narrower than in the designs, where it's 424px.
Fixed

> For some reason case function is not searchable.
Fixed!

ranquild commented on frontend/src/metabase/query_builder/components/expressions/Layout/Layout.module.css:
```diff
@@ -0,0 +1,76 @@
+    minmax(var(--main-width), auto)
+    auto;
+
+  @media (max-width: 920px) {
```
can we use standard breakouts? maybe with mantine?

romeovs replied:
There are not enough standard ones to fill the space between the `md` and `sm` breakpoints here. I've tried solving the issue with screen-relative units and I think it works well enough! (cc @AlessioLaiso)

ranquild commented on frontend/src/metabase/query_builder/components/expressions/Layout/Layout.tsx:
```diff
@@ -0,0 +1,33 @@
+  return <Box {...props} className={cx(S.footer, props.className)} />;
+}
+
+Layout.Header = LayoutHeader;
```
I think we don't normally do that and use components directly. This will make tree-shaking impossible

ranquild commented on frontend/src/metabase/query_builder/components/expressions/ExpressionWidgetHeader.tsx:
```diff
@@ -18,7 +18,7 @@ export const ExpressionWidgetHeader = ({
 }: Props): JSX.Element => {
   return (
-    <Flex className={ExpressionWidgetHeaderS.Header}>
+    <Flex className={ExpressionWidgetHeaderS.Header} maw={688}>
```
magic constant?

ranquild reviewed (commented):
Function order looks arbitrary. @mngr wdyt? 
<img width="520" alt="Screenshot 2025-03-17 at 11 07 19" src="https://github.com/user-attachments/assets/f092eb35-0493-4821-8a74-3903a8e9123a" />

romeovs committed:
Remove unnecessary maw prop

romeovs committed:
Sort functions

mngr commented:
<img width="1143" alt="Captura de ecrã 2025-03-17, às 15 41 07" src="https://github.com/user-attachments/assets/ec96aeb7-f375-4091-8dfc-952c39a43be1" />
@AlessioLaiso Functions browser button when active looks way too prominent, I wonder if we should make it with just a border without the blue background? Also it seems the vertical padding should be bigger between the buttons

romeovs committed:
Add tests for FunctionBrowser

romeovs committed:
Remove nested component definitions

romeovs committed:
Use standard breakpoints

romeovs committed:
Merge branch 'master' into custom-expression-function-browser

romeovs committed:
Add spacing between toolbar buttons

romeovs committed:
Fix functions test

romeovs committed:
Add category to distinct-if

romeovs committed:
Fix tests that were broken by sorting

romeovs committed:
Fix unit test examples

romeovs committed:
Add more unit tests for filtering

romeovs commented:
@ranquild The code compares the [`displayName`](https://github.com/metabase/metabase/pull/55102/files#diff-fd71d097a7ec609ed305b03ef1412cc7197fe53cf4294bd50e4268e0d43dfc14R77) of the clause to the filter. 

So it should match `regexextract` when typing `regexex`, even though the mbql name is `regex-match-first`:

<img width="467" alt="Screenshot 2025-03-18 at 17 15 53" src="https://github.com/user-attachments/assets/61cfe2a3-492e-4b8f-842b-ba3473954510" />


Just to make that extra clear, I've added a unit test to make sure we are matching on `displayName`.

romeovs committed:
Do not enter parentheses when there are no arguments to the clause

romeovs committed:
Sort by displayName

romeovs committed:
Sort by structure

romeovs committed:
Add e2e about clauses with no args

romeovs committed:
Fix sorting test

romeovs commented:
@ranquild oh of course! Fixed and updated the test.

mngr commented:
@romeovs thanks for adding the padding between the buttons! now it looks good
Could you please also check the styles of the function browser, I see a lot of differences compared to the figma 
<img width="665" alt="Captura de ecrã 2025-03-18, às 16 32 42" src="https://github.com/user-attachments/assets/13c5db6a-a99c-4648-9fd7-ab5764760dcb" />

- Borders, icon and text color and size for the search input
- Function names text size (and color?)
- Padding between the functions

romeovs committed:
Fix font sizes for function browser

romeovs committed:
Fix function browser spacing

mngr commented:
> @mngr I've increased the font size and fixed the spacing issues. @AlessioLaiso and I talked about the searchbox and decided to use the Mantine one we currently have instead of implementing a new one.
@romeovs did you commit the changes? cause I don't see them

romeovs commented:
@mngr Please have another look

romeovs committed:
Allow replacing selection with function browser
